File: src/seq_sim.rs
use crate::grid::Grid;
use crate::output;
use crate::rules::next_cell_state;
use std::{thread, time::Duration};

pub fn step(grid: &mut Grid) {
    let new_rust: Vec<Vec<_>> = (0..grid.height)
        .map(|y| (0..grid.width)
            .map(|x| next_cell_state(grid, x, y))
            .collect())
        .collect();

    grid.rust = new_rust;
}

pub fn run(grid: &mut Grid, steps: usize, delay_ms: u64, output_dir: &str) {
    for i in 0..steps {
        clear_terminal();
        println!("Step {}/{}", i + 1, steps);
        grid.display();

        output::save_step(grid, i, output_dir);
        step(grid);

        thread::sleep(Duration::from_millis(delay_ms));
    }

    output::save_step(grid, steps, output_dir);
    output::save_summary(steps, grid.width, grid.height, output_dir);

    clear_terminal();
    println!("Step {}/{} — simulation complete\n", steps, steps);
    println!("Output saved to: {}\n", output_dir);
    grid.display();
}

pub fn step_n(grid: &mut Grid, steps: usize) {
    for _ in 0..steps {
        step(grid);
    }
}

fn clear_terminal() {
    print!("\x1B[2J\x1B[1;1H");
}

File: python/seq_sim.py
import time
import grid
from rules import compute_next_state

def clear():
    print('\033[2J\033[1;1H', end='')

def run(rust, hum, oxy, steps, delay_s=0.1, output_dir='output/python_seq'):
    for i in range(steps):
        clear()
        print(f'Step {i+1}/{steps} [python sequential]')
        grid.display(rust)

        grid.save_step(rust, hum, oxy, i, output_dir)
        rust = compute_next_state(rust, hum, oxy)

        time.sleep(delay_s)

    grid.save_step(rust, hum, oxy, steps, output_dir)
    grid.save_summary(steps, rust.shape[1], rust.shape[0], output_dir)

    clear()
    print(f'Step {steps}/{steps} — sequential complete')
    print(f'Output saved to: {output_dir}\n')
    grid.display(rust)

    return rust

def step_n(rust, hum, oxy, steps):
    for _ in range(steps):
        rust = compute_next_state(rust, hum, oxy)
    return rust